### PR TITLE
Roleselection: Layering

### PR DESCRIPTION
**Pull request by Alf21**
The goal of a layered role-selection is that server-owners can decide the order in which roles are selected. This will fix the issue that detectives often won't be a part of the role-selection because of too many roles that already took place in the selection (automatic displacement).

Currently, it's just a concept:
There is a layered table, that can be modified in the `TTT2ModifyLayeredBaseRoles` hook.
E.g.
```lua
hook.Add("TTT2ModifyLayeredBaseRoles", "LayerDete", function(layeredBaseRolesTbl, availableBaseRolesTbl)
    table.insert(layeredBaseRolesTbl, 1, DETECTIVE)
end)
```
should insert the detective role into the first selection order and should lead to the result, that detectives are selected before other custom roles.
For sure, this is just a concept. Trying to create a layering based on hooks (that currently don't support ordering/priorities) doesn't make a lot of sense. There need to be a built-in VGUI arrangement or similar things.
With this PR, I provided a VGUI based Drag-and-Drop (DnD) solution, accessible with the console command `ttt2_edit_rolelayering`.

Current vgui progress:
grafik

---

### TODO:
- ~Improving the VGUI code/performance~ (Planned with the upcoming VGUI rework)
- ~Add a random var so that there is an option that roles in a layer can have different chances to get selected
     That's a bit weird currently. If setting `ttt_surv_random = 50`, the role will spawn with 25% if being put into a layer with just another role~ (This have to be discussed and isn't necessary for this PR)
- [x] Move `dnd_example.lua` into new file `sh_rolelayering.lua`
- [x] Adding a vertical ~and horizontal~ scrollbar to the DnD-Receiver
- [x] Saving the layers
- [x] Test and probably fix the new role-selection
- [x] Connection between layers and the DnD-VGUI (so that changes take effect)
- [x] Reworking the sub-role layering: They have to be connected with (indexed by) the related base-role
- [x] Add sub-role DnD-VGUI layering (probably selectable with a dropdown)
- [x] Scrollable DnD-Sender
- [x] VGUI DnD-Sender and DnD-Receiver concept
- [x] Sub-roles cleanup
- [x] Layered role-selection
- [x] Converting DnD-Buttons into DnD-Icons (with roles as icons and role names as tooltips)

### CODE REVIEW

File: gamemodes/terrortown/gamemode/server/sv_roleselection.lua [hunk 1]
```diff
@@ -23,6 +25,126 @@ roleselection.cv.ttt_max_roles_pct =  CreateConVar("ttt_max_roles_pct", "0", {FC
...
+			for _, entry in pairs(currentLayerTbl) do -- depth
+				depthCount = depthCount + 1
+				validTbl[baserole][layerCount][depthCount] = entry
+			end
+		end
+	end
+
+	roleselection.subroleLayers = validTbl
+
+	-- TODO remove invalid data from database, but SQL.Delete is missing
```
Comment (Alf21) [inline gamemodes/terrortown/gamemode/server/sv_roleselection.lua:106]: @saibotk :p

File: CHANGELOG.md [hunk 2]
```diff
@@ -8,11 +8,14 @@ All notable changes to TTT2 will be documented here. Inspired by [keep a changel
 
 - Added ConVar to toggle double-click buying
 - Added Japanese translation (by @Westoon)
+- Added Drag&Drop role layering VGUI, accessable with the console command `ttt2_edit_rolelayering`. Keep in mind that `ttt_[roleName]_random` and similar conditions are still taking place in the calculations. 
+  So if you wanna have a real "50/50" selection in a two-entry-sized "OR"-table layer, then you have to set the related `ttt_[roleName]_random` ConVars to 100%.
```
Comment (Histalek) [inline CHANGELOG.md:12]: I get what you are trying to do, but usage/explanations of changes shouldn't be documented in the changelog.
A page in our docs should be the way to go imo

```suggestion
- Added Drag&Drop role layering VGUI, accessible with the console command `ttt2_edit_rolelayering`
```

**Comment by Histalek** [comment]
The review request and the 2 open todos are mildly conflicting i think, you might want to clear that up :)

File: CHANGELOG.md (hunk 2 above)
Comment (TimGoll) [reply CHANGELOG.md:12]: Since this will be removed once integrated into the settings menu, I don't think it is that important

Comment (Alf21) [reply CHANGELOG.md:12]: Yea and the other two ToDos are something that has to be discussed but isn't needed currently / for this PR. Additionally, vgui improvements gonna come with/after the vgui PR.

**Comment by Alf21** [comment]
Maybe we should add the detective to the first layer by default, so the expected "default" TTT behavior is given. What do you think?

**Review by TimGoll — APPROVED** [review]

File: gamemodes/terrortown/gamemode/server/sv_roleselection.lua [hunk 3]
```diff
@@ -394,16 +612,16 @@ end
 -- @note This function modifies the given `plys` var.
 --
 -- @param table plys The players that can receive roles.
--- @param table roleData The @{ROLE} object of the considered role.
+-- @param int subrole The @{ROLE} index of the considered role.
```
Comment (saibotk) [inline gamemodes/terrortown/gamemode/server/sv_roleselection.lua:615]: Is there an `int` type?

File: gamemodes/terrortown/gamemode/server/sv_roleselection.lua [hunk 4]
```diff
@@ -372,17 +590,17 @@ end
 -- @note The subrole replaces a previously selected baserole.
 --
 -- @param table plys The players that should receive roles.
--- @param table roleData The @{ROLE} object of the considered role.
+-- @param int subrole The @{ROLE} index of the considered role.
```
Comment (saibotk) [inline gamemodes/terrortown/gamemode/server/sv_roleselection.lua:593]: Is there an `int` type?

File: gamemodes/terrortown/entities/entities/ttt_logic_role.lua [hunk 5]
```diff
@@ -8,7 +8,7 @@ ENT.Base = "base_point"
 
 local IsValid = IsValid
 
-local ROLE_ANY = 3
+ROLE_ANY = ROLE_ANY or 3
```
Comment (saibotk) [inline gamemodes/terrortown/entities/entities/ttt_logic_role.lua:11]: Is this even needed? Isn't the shared init file loaded before the entity?

File: CHANGELOG.md [hunk 6]
```diff
@@ -10,6 +10,7 @@ All notable changes to TTT2 will be documented here. Inspired by [keep a changel
 
 - Added ConVar to toggle double-click buying
 - Added Japanese translation (by @Westoon)
+- Added Drag&Drop role layering VGUI, accessible with the console command `ttt2_edit_rolelayering`
```
Comment (saibotk) [inline CHANGELOG.md:13]: Please move the changes to the `Unreleased` section again. (Was moved due to a merge/release)

**Review by saibotk — CHANGES_REQUESTED** [review]

File: gamemodes/terrortown/gamemode/server/sv_roleselection.lua (hunk 3 above)
Comment (Alf21) [reply gamemodes/terrortown/gamemode/server/sv_roleselection.lua:615]: https://github.com/TTT-2/TTT2/blame/51be3cb2c75587d9affdaa3a8d8145d3909f5a7f/gamemodes/terrortown/gamemode/server/sv_network_sync.lua#L850, but seems to be wrong, yea.

File: gamemodes/terrortown/entities/entities/ttt_logic_role.lua (hunk 5 above)
Comment (Alf21) [reply gamemodes/terrortown/entities/entities/ttt_logic_role.lua:11]: Nope, tge entity is loaded before the shared file as it seems

File: gamemodes/terrortown/gamemode/server/sv_roleselection.lua (hunk 3 above)
Comment (Alf21) [reply gamemodes/terrortown/gamemode/server/sv_roleselection.lua]: Done with https://github.com/TTT-2/TTT2/pull/649/commits/f26eb2e183979dfd5d55f2660101db70d14b766c

File: gamemodes/terrortown/gamemode/server/sv_roleselection.lua (hunk 4 above)
Comment (Alf21) [reply gamemodes/terrortown/gamemode/server/sv_roleselection.lua]: Done with https://github.com/TTT-2/TTT2/pull/649/commits/f26eb2e183979dfd5d55f2660101db70d14b766c

File: CHANGELOG.md (hunk 6 above)
Comment (Alf21) [reply CHANGELOG.md]: Done with https://github.com/TTT-2/TTT2/pull/649/commits/f26eb2e183979dfd5d55f2660101db70d14b766c

**Review by saibotk — APPROVED** [review]
LGTM

**Pull request merged**